### ISSUE TITLE
Light GBM GOSS Tree working

### ISSUE THREAD

**Issue by jayantphor** [microsoft/LightGBM#3348]
I have implemented  Light Gbm in python but unable to  figure out how the sampling works after 2nd iteration as it takes all the 21 records in first two iterations but it takes only 7 record in 3rd iteration. I have used the following set of parameters to train my model.

 1. **Input Data File-**
[input_data.xlsx](https://github.com/microsoft/LightGBM/files/5163494/input_data.xlsx)

2. **Set of Parameters used for training.** 

x_train=data.iloc[:,1:4]
y_train=data['y']
d_train = lgb.Dataset(x_train, label=y_train)
params['learning_rate'] = 0.4
params['boosting_type'] = 'goss'
params['objective'] = 'binary'
params['metric'] = 'binary_logloss'
params['sub_feature'] = 0.3
params['num_leaves'] = 2
params['min_data'] = 1
params['max_depth'] = -1
params['top_rate'] = 0.2
params['other_rate'] = 0.1
params['min_child_weight'] = 0.1

model = lgb.train(params, d_train, 3
                 )

3. **Output trees for 3 rounds**

![Round1](https://user-images.githubusercontent.com/29829381/92009972-dbe6bb00-ed66-11ea-9a17-4a00b89608fa.png)
![Round2](https://user-images.githubusercontent.com/29829381/92009977-ddb07e80-ed66-11ea-9035-88451cb2a931.png)
![Round3](https://user-images.githubusercontent.com/29829381/92009979-ddb07e80-ed66-11ea-98cb-8f01afe18568.png)

 Can someone help me on the following queries:-
-Which  criteria it is using for sampling 7 records out of 21 in third round(after goss is applied)?  @Laurae2 @jameslamb @StrikerRUS @guolinke @shiyu1994 @Zibi789 
-Which method it is using for sampling records(other_rate*(1-top_rate)*21) with lower gradients?   @Laurae2 
-Where does the weight factor (1-top_rate)/other_rate applied ? @Laurae2

**Comment by guolinke** [comment]
goss is sampling method in gbdt. and it will start to sample after `1.0 / leanring_rate` iterations. as you set learning rate to 0.4, it starts to sample at 3-rd iteration.

**Comment by jayantphor** [comment]
Thanks @guolinke for the answer but how do i know which 7 records it is choosing for third iteration when goss is applied. Is there any method to choose those records?

**Comment by guolinke** [comment]
refer to code https://github.com/microsoft/LightGBM/blob/ca066d499d4b58f74fbd7ab875855cb9f3cb07a1/src/boosting/goss.hpp#L103-L152

basically, it chooses the data with the large gradients (by `top_rate`), and randomly samples the data with small gradients (by `other_rate`).

**Comment by jayantphor** [comment]
Yes i was looking into this code but I have few queries:
-As per documentation it should pick the top k records with larger gradients but why it is taking [gradient* hessian] ?
-For the top_rate(0.2) it is sampling 0.2*21=4.2(~4 records) but on the other side with other_rate(0.1) how it is sampling 0.1*20=2 records out of the remaining 17 records?

**Comment by guolinke** [comment]
@jayantphor 
`hessian` is like the weight of a data sample, so we multiply gradient to it. And the sample weight (if provided) will be multiplied to `hessian` too, so it is useful when using sample weights. 
the other_rate is based on total dataset, not the rest data samples. so it is about `21 * 0.1` samples. and it is sampled randomly: `bagging_rands_[cur_idx / bagging_rand_block_].NextFloat() < prob`.

**Comment by jayantphor** [comment]
Thanks @guolinke.
Can you please tell me how it is choosing a split points(exact value) in case of gbdt and goss ?

**Comment by jayantphor** [comment]
Thanks @guolinke .
I have one more query regarding the factor (1-a)/b which is used inside the GOSS for bumping up the  hessian and gradient value of records with smaller gradients. 
I have 36 records in my input dataset having 5 classes in my target variable  and my top rate and other rate are 0.2 and 0.1 respectively. There can be three ways to calculate the value of this 1-a/b factor i.e,

1. Independent of no of records. i.e, **(1-0.2)/0.1=8** which should be fixed for all set of records
2. Depends on no. of records i.e,  0.2*36=7.2(a=topK) and  0.1*(1-0.2)*36 =2.88 (b= other K)    and hence **(1-a)/b =  (36-7.2)/2.88=10**
3. By rounding of the values of records in **Point 2** i.e, a=7 and b= 3(2.88) and hence **(1-a)/b=(36-7)/3=9.66**

Can anyone of you tell me which is the right approach for calculating value of factor (1-a)/b ? @StrikerRUS @Laurae2 @jameslamb @shiyu1994 @guolinke @Zibi789

**Comment by shiyu1994** [comment]
The code for calculating the factor (1 - a)/b is
> https://github.com/microsoft/LightGBM/blob/1782fcb1e46ad0f06993d403b62d013d1e2cf76b/src/boosting/goss.hpp#L114-L120

It is basically the same as your Point 3, except that
1. The `other_rate` is multiplied with the size of whole dataset (`cnt`), instead of the remaining dataset after sampling the top instances.
2. `static_cast<data_size_t>` is a rounding down operation for positive floating numbers, instead of rounding to the nearest integer.

**Comment by jayantphor** [comment]
Thanks @shiyu1994 .
So according to this code , for 36 records my (1-a)/b factor should be 9(after rounded off).
